### DISCUSSION TITLE
Ability to set column width and `className` for <th> and <td> in shadcn-ui Data Table

### DISCUSSION THREAD

**Discussion by TheBassGuy** [shadcn-ui/ui#1122] [Q&A] [answered]
Hi, I am unable to add classNames to <th> and <td> in shadcn-ui Data Table, I used the same method as used on [Data Table - shadcn/ui](https://ui.shadcn.com/docs/components/data-table#basic-table) but I am unable to set classNames in the above stated elements

**Comment by TheBassGuy** [comment]
I want to do something like this
```
export const columns: ColumnDef<Product>[] = [
    {
        accessorKey: "column1",
        header: "Column 1",
        headClassName : 'w-16 bg-yellow-400 hidden lg:block',
        cellClassName : 'font-bold'
    },
    {
        accessorKey: "column2",
        header: "Column 2",
        headClassName : 'w-32 bg-green-400 hidden lg:block',
        cellClassName : 'italic'
    },
]

**Comment by corysimmons** [comment]
I copied and pasted this snippet into VSCode:
![image](https://github.com/shadcn-ui/ui/assets/224636/23d7769e-7d95-4609-a3bb-824abdbddc1c)

Then I looked for `className`:
![image](https://github.com/shadcn-ui/ui/assets/224636/158d1bdc-23dc-4451-9cb1-1476e6baaf6d)

I think you should be able to target and tweak the header row (to include a `className`) in a similar fashion. Notice how Shadcn has rewritten the HTML of the header to include a checkbox here:
![image](https://github.com/shadcn-ui/ui/assets/224636/49498098-3d03-4ccc-8b91-69b365838db0)

All this to say, you're going about this the wrong way. The `columns` array of objects doesn't have any `headClassName` or `cellClassName` and likely never will (since that array of objects is just a configuration for [Tanstack Table](https://tanstack.com/table/v8)).

But you _can_ rewrite the markup completely and use things like `getValue` for all types of cells (headers and otherwise).

Hope this helps. If you make a CodeSandbox or something, I'll be happy to tweak it to work for you.

**Reply by TheBassGuy** [reply]
Hi, thanks for the reply, I tried using this, but this is editing the contents inside the <th> or <td> rather than the <th> or <td>, I wanted to edit the <th> so that I can set the column widths and maybe hide some columns on smaller viewports...

Please lmk if I can do something for the same...

**Reply by corysimmons** [reply]
I think you're going to have to write custom markup. Instead of destructuring `({ table })`, maybe just `(a)` then `console.log(a)` to see what is available.

Then use some media query plugin like react-responsive to change some state like `visibleColumnIndexes: [number]`.

If you draw a picture of what you're trying to do and create a CodeSandbox with React + shadcn-ui installed and working, I'll try to tinker around and figure it out.

**Reply by TheBassGuy** [reply, marked as answer]
Hi, just found a workaround, added an array of classNames as a prop in data-table.tsx, and took the column index from the map in data-table.tsx, now I just pass the classNames like `["w-32", "w-48", "", "w-12", "w-32", "w-16"]`, works fine for me, thanks for the help

This discussion can be hence closed now

Thanks everyone for making such a beautiful library :D

**Comment by DiegoxK** [comment]
I found that you can actually pass any property you want to the columns definitions

If you put a custom atribute like `headClassName: "w-[80px]"`, it gets passed to the `header.column.columnDef` object when you map through the `headerGroup.headers`:

![image](https://github.com/shadcn-ui/ui/assets/86527054/3d921660-5fbe-446c-8c0f-b02b81a5e64b)

Then you can put that `headClassName` inside the `<TableHead>` `ClassName`.

The only problem was extending the type definitions for `ColumnDef` to tell typescript that my custom `headClassName` was a valid property.

Here is a workaround I glued with help of AI, though I don't think this is the cleanest way to do it:

**definitions.ts**

```typescript
import type { ColumnDef } from "@tanstack/react-table";

export type CustomColumnDef<TData, TValue> = ColumnDef<TData, TValue> & {
  headClassName?: string;
};
```
**columns.tsx**
```typescript
import type { CustomColumnDef } from "@/lib/definitions";

export const columns: CustomColumnDef<Todo, unknown>[] = [
  {
    id: "actions",
    headClassName: "w-[80px]",
    header: "Actions",
  },
];
```

**data-table.tsx**
```typescript
import type { CustomColumnDef } from "@/lib/definitions";

interface DataTableProps<TData, TValue> {
  columns: CustomColumnDef<TData, TValue>[];
  data: TData[];
}

export function DataTable<TData, TValue>({
  columns,
  data,
}: DataTableProps<TData, TValue>) {
  ....

  return (
       ...
          <TableHeader>
            {table.getHeaderGroups().map((headerGroup) => (
              <TableRow key={headerGroup.id}>
                {headerGroup.headers.map((header) => {
                  const headClassName = (
                    header.column.columnDef as CustomColumnDef<unknown, unknown>
                  ).headClassName;
                  return (
                    <TableHead className={headClassName} key={header.id}>
                      {header.isPlaceholder
                        ? null
                        : flexRender(
                            header.column.columnDef.header,
                            header.getContext(),
                          )}
                    </TableHead>
                  );
                })}
              </TableRow>
            ))}
          </TableHeader>
       ...
  );
}

```

It manages to prevent typescript from complaining but I'm not certain if the `ColumnDef` type can be modified in a way that allows you to easily pass custom props

**Reply by DiegoxK** [reply]
Update: You can use the meta property to add any custom property you want
[https://stackoverflow.com/a/77866912](url)
